states/huntsman/calibrating.py

```diff
@@ -4,116 +4,87 @@
+            narrow_band_cameras.append(cam_name)
+        else:
+            broad_band_cameras.append(cam_name)
+
+    return narrow_band_cameras, broad_band_cameras
+
 
+def on_enter(event_data):
+    '''
+    Calibrating state. If safe to do so, take flats and darks. Should be
+    called once at the beginning and end of the night.
+    '''
+    pocs = event_data.model
     pocs.next_state = 'parking'
 
-    try:
+    # Leave the state if not safe to take calibrations
+    if not pocs.is_safe(horizon='flat'):
+        print('Exiting calibrating state because it is no longer safe.')
+        return
 
-        if pocs.observatory.take_flat_fields:
-
-            # Wait for twilight if needed
-            while True:
-                sun_pos = pocs.observatory.observer.altaz(
-                    current_time(),
-                    target=get_sun(current_time())
-                ).alt
-
-                if sun_pos.value <= 10 and sun_pos.value >= 0:
-                    pocs.say("Sun is still not down yet, will wait to take some flats")
-                    pocs.sleep(delay=60)
-                else:
-                    break
-
-            # Take the flats
-            if sun_pos.value <= 0 and sun_pos.value >= -12:
-                pocs.say("Taking some flat fields to start the night")
-
-                narrow_band_cameras = list()
-                broad_band_cameras = list()
-                for cam_name, cam in pocs.observatory.cameras.items():
-                    if cam.filter_type.lower().startswith('ha'):
-                        narrow_band_cameras.append(cam_name)
-                    else:
-                        broad_band_cameras.append(cam_name)
-
-                if len(narrow_band_cameras) > 0:
-                    pocs.say("Starting narrow band flat fields")
-                    pocs.observatory.take_evening_flats(camera_list=narrow_band_cameras)  # H-alpha
-
-                if len(broad_band_cameras) > 0:
-                    pocs.say("Staring broad band flat fields")
-                    pocs.observatory.take_evening_flats(camera_list=broad_band_cameras)   # g and r
+    if pocs.observatory.take_flat_fields:
 
-    except Exception as e:
-        pocs.logger.warning("Problem with flat fielding: {}".format(e))
+        try:
+            # Identify filter types
+            narrow_band_cameras, broad_band_cameras = _get_cameras(pocs)
+
+            # Take calibration frames
+            pocs.say("Starting narrow band flat fields")
+            pocs.observatory.take_evening_flats(camera_list=narrow_band_cameras)
+
+            pocs.say("Staring broad band flat fields")
+            pocs.observatory.take_evening_flats(camera_list=broad_band_cameras)
+
+        except Exception as err:
+            pocs.logger.warning(f'Problem with flat fielding: {err}')
+
+    # This code needs to be moved to its own state
+    # Wait for twilight
+    coarse_focus_timeout = 600  # Put this in a config file
```

Comment:
Can you move this line below the safety check and grouped with the focus events?

Reply:
I will be moving all this in an imminent PR so no need to worry for now.